api/comments_api.py

```diff
@@ -22,13 +23,21 @@
 from internals import notifier
 
 
-def comment_to_json_dict(comment):
+def comment_to_json_dict(comment, user_email=None, show_deleted=False):
+  content = comment.content
+  # If the comment is deleted and the current user can't view it,
+  # don't send the comment content to the client.
+  if (comment.deleted_by is not None and
+      user_email != comment.deleted_by and not show_deleted):
+    content = "[Deleted]"
```

Comment:
Thanks for including he screenshots, that makes it much easier to think about the UX in addition to reviewing the code.

I guess I didn't think of this before, but now I think that the last screenshot you posted (the case where a comment is deleted and the user lacks perms) should just not be in the API response at all.  That could be done before line 57 by filtering the list of comments before converting them to json_dicts.  The filtering could use the same condition that you use here.

Reply:
Good idea, I reverted the json_dicts function and added a filter beforehand!